content/asm_2.md

```diff
@@ -1,235 +1,508 @@
+
+```C
+int foo(int arg1, int arg2, int arg3, int arg4, int arg5, int arg6, int arg7, int arg8) {
+    return arg1 + arg2 + arg3 + arg4 + arg5 + arg6 + arg7 + arg8;
+}
+
+int bar() {
+    return foo(1, 2, 3, 4, 5, 6, 7, 8);
+}
+```
+
+If we compile it and look at the assembly code, we will see such an output of the program:
+
+```assembly
+bar:
+        ;; Preserve the base pointer
+        push    rbp
+        ;; Preserve the stack pointer
+        mov     rbp, rsp
+        ;; Push the eight argument on the stack
+        push    8
+        ;; Push the seventh argument on the stack
+        push    7
+        ;; Push the sixth argument on the stack
+        mov     r9d, 6
+        ;; Push the fifth argument on the stack
+        mov     r8d, 5
+        ;; Push the fourth argument on the stack
+        mov     ecx, 4
+        ;; Push the third argument on the stack
+        mov     edx, 3
+        ;; Push the second argument on the stack
+        mov     esi, 2
+        ;; Push the first argument on the stack
+        mov     edi, 1
+        ;; Call the function `foo`
+        call    foo
+        ;; Clean-up the stack from the 8th and 7th arguments
+        add     rsp, 16
+        ;; Restore the old rbp
+        leave
+        ;; Return from the function
+        ret
+foo:
+        ;; Preserve the base pointer
+        push    rbp
+        ;; Preserve the stack pointer
+        mov     rbp, rsp
+        ;; Move 4 bytes value from the edi register to the address stored in the rbp register minus 4 bytes offset
+        mov     DWORD PTR [rbp-4], edi
+        ;; Move 4 bytes value from the esi register to the address stored in the rbp register minus 8 bytes offset
+        mov     DWORD PTR [rbp-8], esi
+        ;; Move 4 bytes value from the edx register to the address stored in the rbp register minus 12 bytes offset
+        mov     DWORD PTR [rbp-12], edx
+        ;; Move 4 bytes value from the ecx register to the address stored in the rbp register minus 16 bytes offset
+        mov     DWORD PTR [rbp-16], ecx
+        ;; Move 4 bytes value from the r8d register to the address stored in the rbp register minus 20 bytes offset
+        mov     DWORD PTR [rbp-20], r8d
+        ;; Move 4 bytes value from the r9d register to the address stored in the rbp register minus 24 bytes offset
+        mov     DWORD PTR [rbp-24], r9d
+        ...
+        ... # Skip arithmetic operations for now
+        ...
+        ;; Restore the old rbp
+        pop     rbp
+        ;; Return from the function
+        ret
+```
+
+> [!NOTE]
+> The C program should be compiled without any optimization flags because the compiler may eliminate all the calculation and just calculate the sum of arguments in compile time. You can use `-O0 -masm=intel` flags for compiler to avoid optimization. You can use tools like [godbolt](https://godbolt.org/) to see the assembly output of these functions.
```

Comment:
```suggestion
> The C program should be compiled without any optimization flags because the compiler may eliminate all the calculations and just calculate the sum of arguments in compile time. You can use `-O0 -masm=intel` flags for the compiler to avoid optimization. You can use tools like [godbolt](https://godbolt.org/) to see the assembly output of these functions.
```